File: scripts/build-seo-pages.js
const fs = require('fs');
const path = require('path');

const rootDir = __dirname + '/..';
const indexHtml = fs.readFileSync(path.join(rootDir, 'index.html'), 'utf-8');

const TOOLS = {
  merge: { title: 'Merge PDF', desc: 'Combine unlimited PDF files — no restrictions' },
  split: { title: 'Split PDF', desc: 'Extract pages or split into separate files' },
  compress: { title: 'Compress PDF', desc: 'Reduce file size while preserving text and vector quality' },
  jpg2pdf: { title: 'JPG to PDF', desc: 'Convert images to PDF — EXIF rotation corrected automatically' },
  pdf2jpg: { title: 'PDF to JPG', desc: 'Extract pages as high-quality JPG or PNG images' },
  redact: { title: 'Cover Area', desc: 'Hide watermarks, signatures or sensitive data with an opaque cover' },
  rotate: { title: 'Rotate PDF', desc: 'Fix page orientation in any PDF' },
  extract: { title: 'Extract Pages', desc: 'Pull selected pages into a new PDF — with smart presets' },
  watermark: { title: 'Watermark PDF', desc: 'Add text watermark to every page — diagonal, tiled or positioned' },
  pagenum: { title: 'Add Page Numbers', desc: 'Number pages — Arabic, Roman or alphabetic, any position' },
  meta: { title: 'Edit Metadata', desc: 'View and edit PDF title, author, subject and other fields' },
  protect: { title: 'Protect PDF', desc: 'Add open password & restrict permissions — AES-256, fully client-side' },
};

// Also create 404.html for GitHub Pages fallback
const fallbackHtml = indexHtml.replace(
  '<head>',
  `<head>
  <script>
    // GitHub Pages 404 fallback to route correctly
    (function(){
      const l = window.location;
      if (l.pathname.endsWith('-pdf/') || l.pathname.endsWith('-pdf') || l.pathname.endsWith('2pdf/') || l.pathname.endsWith('2jpg/')) {
        sessionStorage.redirect = l.href;
      }
    })();
  </script>`
);
fs.writeFileSync(path.join(rootDir, '404.html'), fallbackHtml);

for (const [tool, meta] of Object.entries(TOOLS)) {
  const folderName = tool.includes('pdf') ? tool : `${tool}-pdf`;
  const toolDir = path.join(rootDir, folderName);
  if (!fs.existsSync(toolDir)) {
    fs.mkdirSync(toolDir);
  }

  // Update relative paths to point to parent
  let html = indexHtml
    .replace(/(href|src)="css\//g, '$1="../css/')
    .replace(/(href|src)="js\//g, '$1="../js/')
    .replace(/(href|src)="img\//g, '$1="../img/')
    // Handle inline script injection
    .replace('<title>PDFree — Process PDF locally. Free &amp; Private</title>', `<title>${meta.title} — PDFree</title>`)
    .replace('<meta name="description" content="Merge, split, compress, protect and watermark PDF files securely in your browser. 100% free, no data leaves your device.">', `<meta name="description" content="${meta.desc}">`)
    .replace('</body>', `
  <script>
    // Pre-select tool based on directory name
    window.PDFREE_INITIAL_TOOL = '${tool}';
  </script>
</body>`);

  fs.writeFileSync(path.join(toolDir, 'index.html'), html);
  console.log(`Generated ${folderName}/index.html`);
}

File: scripts/build-seo-pages.py
import os
import json
from jinja2 import Environment, FileSystemLoader

# Configuration
DOMAIN = "https://mahmudovbahrom555-lab.github.io/PDFree"
LANGUAGES = ['en', 'es', 'pt']
DEFAULT_LANG = 'en'

root_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
scripts_dir = os.path.join(root_dir, 'scripts')

# Load Translations
with open(os.path.join(scripts_dir, 'translations.json'), 'r', encoding='utf-8') as f:
    translations = json.load(f)

# Jinja2 Environment
env = Environment(loader=FileSystemLoader(root_dir))
template = env.get_template('index.template.html')

def t(key, lang):
    """Retrieve translation with fallback to DEFAULT_LANG"""
    parts = key.split('.')
    category = parts[0]
    sub_key = '.'.join(parts[1:])
    
    try:
        val = translations[category][sub_key].get(lang)
        if not val:
            val = translations[category][sub_key].get(DEFAULT_LANG)
        return val
    except KeyError:
        return f"[{key}]"

def url_helper(tool_key, lang):
    """Generate the relative path for a tool based on language"""
    slug = t(f"seo.{tool_key}.slug", lang)
    if lang == DEFAULT_LANG:
        # e.g., merge-pdf/
        folder = slug if 'pdf' in slug else f"{slug}-pdf"
        return f"{folder}/"
    else:
        # e.g., es/unir-pdf/
        folder = slug if 'pdf' in slug else f"{slug}-pdf"
        return f"{lang}/{folder}/"

def absolute_url(tool_key, lang):
    """Generate the absolute URL for a tool based on language"""
    slug = t(f"seo.{tool_key}.slug", lang)
    folder = slug if 'pdf' in slug else f"{slug}-pdf"
    rel_path = f"{folder}/" if lang == DEFAULT_LANG else f"{lang}/{folder}/"
    return f"{DOMAIN}/{rel_path}"

tools = ['merge', 'split', 'compress', 'jpg2pdf', 'pdf2jpg', 'extract', 'watermark', 'pagenum', 'meta', 'redact', 'rotate', 'protect']
sitemap_urls = []

# Generate Pages
for lang in LANGUAGES:
    # Set the base directory for the language
    lang_dir = root_dir if lang == DEFAULT_LANG else os.path.join(root_dir, lang)
    if not os.path.exists(lang_dir):
        os.makedirs(lang_dir)

    # 1. Generate language homepage (redirect to default tool, usually merge)
    # The actual HTML is just the merge tool for now, matching the old behavior
    # But wait, PDFree root index.html is essentially the merge tool.
    # We will generate it exactly like the merge tool, but placed at lang_dir/index.html
    
    for tool in tools:
        # Determine output path
        slug = t(f"seo.{tool}.slug", lang)
        folder_name = slug if 'pdf' in slug else f"{slug}-pdf"
        
        # If default lang, folder is root_dir / folder_name
        # If other lang, folder is root_dir / lang / folder_name
        tool_dir = os.path.join(lang_dir, folder_name)
        if not os.path.exists(tool_dir):
            os.makedirs(tool_dir)

        # Generate hreflang tags
        hreflang_tags = ""
        for l in LANGUAGES:
            hreflang_tags += f'  <link rel="alternate" hreflang="{l}" href="{absolute_url(tool, l)}">\n'
        hreflang_tags += f'  <link rel="alternate" hreflang="x-default" href="{absolute_url(tool, DEFAULT_LANG)}">'

        # Determine relative paths to assets
        # If DEFAULT_LANG: inside /merge-pdf/, depth is 1
        # If other lang: inside /es/unir-pdf/, depth is 2
        base_prefix = "../" if lang == DEFAULT_LANG else "../../"

        # Context for Jinja
        context = {
            'lang': lang,
            'canonical_url': absolute_url(tool, lang),
            'hreflang_tags': hreflang_tags,
            'base_url': base_prefix if lang == DEFAULT_LANG else f"{base_prefix}{lang}/",
            'current_tool': tool,
            't': lambda k: t(k, lang),
            'url': lambda k: f"{base_prefix}{t(f'seo.{k}.slug', lang)}{'-pdf' if 'pdf' not in t(f'seo.{k}.slug', lang) else ''}/" if lang == DEFAULT_LANG else f"{base_prefix}{lang}/{t(f'seo.{k}.slug', lang)}{'-pdf' if 'pdf' not in t(f'seo.{k}.slug', lang) else ''}/",
        }

        # Render HTML
        html = template.render(context)
        
        # Fix asset paths dynamically since they are deep
        # A simpler way is to replace standard asset prefixes
        html = html.replace('href="css/', f'href="{base_prefix}css/')
        html = html.replace('src="js/', f'src="{base_prefix}js/')
        html = html.replace('href="manifest.json"', f'href="{base_prefix}manifest.json"')
        html = html.replace('href="icons/', f'href="{base_prefix}icons/')
        
        # Write file
        with open(os.path.join(tool_dir, 'index.html'), 'w', encoding='utf-8') as f:
            f.write(html)
            
        sitemap_urls.append(absolute_url(tool, lang))

        # Root index.html should be a copy of the merge tool for the respective language
        if tool == 'merge':
            # Root needs depth adjusted
            root_context = context.copy()
            root_context['base_url'] = "./" if lang == DEFAULT_LANG else f"../{lang}/"
            root_html = template.render(root_context)
            # No base_prefix replacement needed for DEFAULT_LANG root, 
            # but for /es/index.html depth is 1.
            root_base_prefix = "" if lang == DEFAULT_LANG else "../"
            if root_base_prefix:
                root_html = root_html.replace('href="css/', f'href="{root_base_prefix}css/')
                root_html = root_html.replace('src="js/', f'src="{root_base_prefix}js/')
                root_html = root_html.replace('href="manifest.json"', f'href="{root_base_prefix}manifest.json"')
                root_html = root_html.replace('href="icons/', f'href="{root_base_prefix}icons/')
            
            with open(os.path.join(lang_dir, 'index.html'), 'w', encoding='utf-8') as f:
                f.write(root_html)
            
            sitemap_urls.append(f"{DOMAIN}/" if lang == DEFAULT_LANG else f"{DOMAIN}/{lang}/")

print("Generated all localized SEO pages successfully.")

# Generate sitemap.xml
import datetime
today = datetime.datetime.now().strftime("%Y-%m-%d")

sitemap_content = '<?xml version="1.0" encoding="UTF-8"?>\n'
sitemap_content += '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">\n'

for url in set(sitemap_urls): # Use set to avoid duplicates
    sitemap_content += '  <url>\n'
    sitemap_content += f'    <loc>{url}</loc>\n'
    sitemap_content += f'    <lastmod>{today}</lastmod>\n'
    sitemap_content += '    <changefreq>weekly</changefreq>\n'
    sitemap_content += '    <priority>0.8</priority>\n'
    sitemap_content += '  </url>\n'

sitemap_content += '</urlset>'

with open(os.path.join(root_dir, 'sitemap.xml'), 'w', encoding='utf-8') as f:
    f.write(sitemap_content)

print("Generated sitemap.xml")

# Fix 404.html using the template
fallback_context = {
    'lang': DEFAULT_LANG,
    'canonical_url': f"{DOMAIN}/",
    'hreflang_tags': '',
    'base_url': "./",
    'current_tool': 'merge',
    't': lambda k: t(k, DEFAULT_LANG),
    'url': lambda k: f"./{t(f'seo.{k}.slug', DEFAULT_LANG)}{'-pdf' if 'pdf' not in t(f'seo.{k}.slug', DEFAULT_LANG) else ''}/",
}
fallback_html = template.render(fallback_context)
fallback_html = fallback_html.replace(
    '<head>',
    '''<head>
  <script>
    // GitHub Pages 404 fallback to route correctly
    (function(){
      const l = window.location;
      if (l.pathname.endsWith('-pdf/') || l.pathname.endsWith('-pdf') || l.pathname.endsWith('2pdf/') || l.pathname.endsWith('2jpg/')) {
        sessionStorage.redirect = l.href;
      }
    })();
  </script>'''
)
with open(os.path.join(root_dir, '404.html'), 'w', encoding='utf-8') as f:
    f.write(fallback_html)
